content/posts/kong-dynamic-plugin-ordering.adoc

```diff
@@ -0,0 +1,170 @@
+---
+title: "Kong Dynamic Plugin Ordering"
+description: "This is a tutorial how you can use a new dynamic plugin ordering in Kong Gateway."
+date: 2022-09-09
+draft: false
+slug: "plugin-ordering"
+authors: ["Viktor Gamov"]
+tags: ["Kong", "tutorial"]
+categories: ["Kong", "Gateway"]
+externalLink : ""
+series: []
+---
+:source-highlighter: rouge
+:imagesdir: /images/2022/
+:rouge-style: base16
+:icons: font
+:toc: macro
+
+.TL;DR
+NOTE: Use a new dynamic plugin ordering to control plugins' execution explicitly.
+You can read more in the official documentation.
+TODO add documentation links.
+
+In this blog post/tutorial, you will learn:
+
+- Why plugin ordering is essential.
+- How you can use dynamic plugin ordering.
+
+WARNING: This feature is only available in Kong Enterprise starting from version 3.0
+
+<<<
+
+// - 1st paragraph: The problem and recommendation summary. The background can be in an appendix.
+// - 2nd paragraph: Further explanation of the situation.
+// - 3rd paragraph: The root cause and solution. Other options considered can be in an appendix.
+// - 4th paragraph: Full recommendation and implementation plan.
+
+toc::[]
+
+== Introduction of the Problem.
+
+All the bundled Kong Gateway plugins have static execution priority.
+It means the `pre-function` plugin, for example, will always be running before `bot-detection` or `zipkin` plugins, `mtls-auth` will always be running before `oauth2`.
+And when customers have a requirement to customize execution order (like run `rate-limiting` before `key-auth`) they can't do that easily.
+
+Projects like https://github.com/Kong/priority-updater[`priority-updater`] can wrap existing plugins and update execution order.
+But this would be considered a hack and workaround for the actual problem.
+
+In Kong Gateway 3.0 we're introducing a _dynamic plugin ordering_ functionality.
+You can specify a directed graph of plugins.
+With this, we should be able to specify config that `rate-limiting` should run before `key-auth` or `oath2-introspection`.
+
+== Example: Default behavior before Kong 3.0
+
+.I have Kong Gateway 3.0 Enterprise instance running locally.
+image::kong_gw_3_0_insomnia.png[]
```

Comment:
It will be a bit confusing to tell the reader "behavior before Kong 3.0", but then show them an image and text with 3.0